src/java/org/apache/cassandra/index/sai/StorageAttachedIndexBuilder.java

```diff
@@ -354,8 +353,16 @@ private void completeSSTable(StorageAttachedIndexWriter indexWriter,
         sstable.registerComponents(group.activeComponents(sstable), tracker);
         if (!replacedComponents.isEmpty())
             sstable.unregisterComponents(replacedComponents, tracker);
-        Set<StorageAttachedIndex> incomplete = group.onSSTableChanged(Collections.emptyList(), Collections.singleton(sstable), existing, false);
 
+        /**
+         * During memtable flush, it completes the transaction first which opens the flushed sstable,
+         * and then notify the new sstable to SAI. Here we should do the same.
+         */
+        txn.trackNewAttachedIndexFiles(sstable);
+        // there is nothing to commit. Close() effectively abort the transaction.
+        txn.close();
```

Comment:
this close() was not called before this patch, where was it used to be called ?
we are closing an object that has been passed as parameter, is the any side effect ?

Reply:
> this close() was not called before this patch, where was it used to be called ?

it's try-resource in the outer method, `close()` is called automatically